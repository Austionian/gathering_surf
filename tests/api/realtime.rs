use gathering_surf::ATWATER_REALTIME_PATH;
use wiremock::{
    matchers::{method, path},
    Mock, ResponseTemplate,
};

use crate::helpers::{start_integration_test_app, start_test_app};

#[tokio::test]
async fn integration_it_returns_the_latest_data_as_json() {
    let app = start_integration_test_app()
        .await
        .expect("Unable to start test server.");

    let response = reqwest::get(format!("http://{}/api/realtime", &app.addr))
        .await
        .unwrap();

    assert_eq!(response.status().as_u16(), 200);

    let data = response.text().await.unwrap();

    assert!(data.contains("as_of"));
    assert!(data.contains("wind_direction"));
    assert!(data.contains("air_temp"));
    assert!(data.contains("quality_color"));
    assert!(data.contains("quality_text"));
}

#[tokio::test]
async fn it_handles_a_non_200_response_from_realtime_client_and_retries_once() {
    let app = start_test_app()
        .await
        .expect("Unable to start test server.");

    let mock_client = &app.mock_client.unwrap();
    Mock::given(method("GET"))
        .and(path(ATWATER_REALTIME_PATH))
        .respond_with(ResponseTemplate::new(502).set_body_string("Bad gateway"))
        .expect(2)
        .mount(mock_client)
        .await;

    let response = reqwest::get(format!("http://{}/api/realtime", &app.addr))
        .await
        .unwrap();

    assert_eq!(response.status().as_u16(), 500);

    let data = response.text().await.unwrap();

    assert!(!data.contains("as_of"));
    assert!(data.contains("Something went wrong: Non 200 response from NOAA realtime"));
}
