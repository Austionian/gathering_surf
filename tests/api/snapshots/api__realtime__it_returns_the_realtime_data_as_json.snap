---
source: tests/api/realtime.rs
expression: data
---
{"as_of":"Fri, 23 May 2025 13:30:00","wind_direction":90,"wind_speed":"4","gusts":"7","water_temp":"46","air_temp":"46","quality_color":"#0bd674","quality_text":"Good","wave_height":"0.98","wave_period":5,"wave_direction":204,"loaded_from_fallback":false}
